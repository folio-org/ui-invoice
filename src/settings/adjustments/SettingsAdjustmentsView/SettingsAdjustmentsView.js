import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { get } from 'lodash';

import {
  Button,
  Checkbox,
  Col,
  ConfirmationModal,
  Icon,
  KeyValue,
  Layer,
  Pane,
  Row,
} from '@folio/stripes/components';
import { ViewMetaData } from '@folio/stripes/smart-components';

class SettingsAdjustmentsView extends Component {
  static propTypes = {
    close: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    adjustment: PropTypes.object,
    rootPath: PropTypes.string,
  };

  static defaultProps = {
    adjustment: { adjustment: {} },
  }

  state = {
    showConfirmDelete: false,
  };

  deleteAdjustment = () => {
    const { onDelete } = this.props;

    this.hideConfirmDelete();
    onDelete();
  };

  showConfirmDelete = () => this.setState({ showConfirmDelete: true });

  hideConfirmDelete = () => this.setState({ showConfirmDelete: false });

  getActionMenu = ({ onToggle }) => {
    const { rootPath, adjustment } = this.props;
    const id = get(adjustment, 'id');

    return (
      <div data-test-view-adjustment-actions>
        <Button
          data-test-view-adjustment-action-edit
          buttonStyle="dropdownItem"
          to={`${rootPath}/${id}/edit`}
        >
          <Icon icon="edit">
            <FormattedMessage id="ui-invoice.button.edit" />
          </Icon>
        </Button>
        <Button
          data-test-view-adjustment-action-delete
          buttonStyle="dropdownItem"
          onClick={() => {
            onToggle();
            this.showConfirmDelete();
          }}
        >
          <Icon icon="trash">
            <FormattedMessage id="ui-invoice.button.delete" />
          </Icon>
        </Button>
      </div>
    );
  };

  render() {
    const {
      close,
      adjustment,
    } = this.props;
    const { showConfirmDelete } = this.state;
    const {
      metadata,
      title,
      adjustment: { alwaysShow, defaultAmount, description, type, prorate, relationToTotal },
    } = adjustment;

    return (
      <Layer
        contentLabel="Adjustment details"
        isOpen
      >
        <Pane
          actionMenu={this.getActionMenu}
          id="invoice-settings-adjustment-view"
          defaultWidth="fill"
          paneTitle={title}
          dismissible
          onClose={close}
        >
          <Row center="xs">
            <Col xs={12} md={8}>
              <Row start="xs">
                <Col xs={12}>
                  {metadata && <ViewMetaData metadata={metadata} />}
                </Col>
                <Col
                  xs={3}
                  data-test-description
                >
                  <KeyValue
                    label={<FormattedMessage id="ui-invoice.settings.adjustments.description" />}
                    value={description}
                  />
                </Col>
                <Col
                  xs={3}
                  data-test-type
                >
                  <KeyValue
                    label={<FormattedMessage id="ui-invoice.settings.adjustments.type" />}
                    value={type}
                  />
                </Col>
                <Col
                  data-test-always-show
                  xs={3}
                >
                  <KeyValue label={<FormattedMessage id="ui-invoice.settings.adjustments.alwaysShow" />}>
                    <Checkbox
                      checked={alwaysShow}
                      disabled
                      type="checkbox"
                    />
                  </KeyValue>
                </Col>
                <Col
                  data-test-default-amount
                  xs={3}
                >
                  <KeyValue
                    label={<FormattedMessage id="ui-invoice.settings.adjustments.defaultAmount" />}
                    value={defaultAmount}
                  />
                </Col>
                <Col
                  data-test-prorate
                  xs={3}
                >
                  <KeyValue
                    label={<FormattedMessage id="ui-invoice.settings.adjustments.prorate" />}
                    value={prorate}
                  />
                </Col>
                <Col
                  data-test-relation-to-total
                  xs={3}
                >
                  <KeyValue
                    label={<FormattedMessage id="ui-invoice.settings.adjustments.relationToTotal" />}
                    value={relationToTotal}
                  />
                </Col>
              </Row>
            </Col>
          </Row>
          {showConfirmDelete && (
            <ConfirmationModal
              id="delete-adjustment-modal"
              confirmLabel={<FormattedMessage id="ui-invoice.settings.adjustments.confirmDelete.confirmLabel" />}
              heading={<FormattedMessage id="ui-invoice.settings.adjustments.confirmDelete.heading" values={{ description }} />}
              message={<FormattedMessage id="ui-invoice.settings.adjustments.confirmDelete.message" />}
              onCancel={this.hideConfirmDelete}
              onConfirm={this.deleteAdjustment}
              open
            />
          )}
        </Pane>
      </Layer>
    );
  }
}

export default SettingsAdjustmentsView;
