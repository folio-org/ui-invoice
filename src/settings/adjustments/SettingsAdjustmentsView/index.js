export { default as SettingsAdjustmentsView } from './SettingsAdjustmentsView';
export { default } from './SettingsAdjustmentsViewContainer';
