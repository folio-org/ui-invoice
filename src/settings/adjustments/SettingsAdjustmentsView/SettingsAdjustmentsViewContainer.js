import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import { FormattedMessage } from 'react-intl';

import { get } from 'lodash';

import { Callout } from '@folio/stripes/components';
import { stripesConnect } from '@folio/stripes/core';

import { CONFIG_ADJUSTMENT } from '../../../common/resources';
import { getSettingsAdjustmentsList } from '../util';
import SettingsAdjustmentsView from './SettingsAdjustmentsView';

class SettingsAdjustmentsViewContainer extends Component {
  static manifest = Object.freeze({
    configAdjustment: CONFIG_ADJUSTMENT,
  });

  static propTypes = {
    close: PropTypes.func.isRequired,
    mutator: PropTypes.object.isRequired,
    match: ReactRouterPropTypes.match.isRequired,
    rootPath: PropTypes.string.isRequired,
    showSuccessDeleteMessage: PropTypes.func.isRequired,
    resources: PropTypes.object,
  };

  createCalloutRef = ref => {
    this.callout = ref;
  };

  deleteAdjustment = async () => {
    const { close, mutator: { configAdjustment }, match: { params: { id } }, showSuccessDeleteMessage } = this.props;

    try {
      await configAdjustment.DELETE({ id });
      close();
      showSuccessDeleteMessage();
    } catch (e) {
      this.callout.sendCallout({
        message: <FormattedMessage id="ui-invoice.settings.adjustments.remove.error" />,
        type: 'error',
      });
    }
  };

  render() {
    const { close, resources, rootPath } = this.props;
    const adjustments = getSettingsAdjustmentsList(get(resources, 'configAdjustment.records', []));
    const adjustment = get(adjustments, '0');

    return (
      <Fragment>
        <SettingsAdjustmentsView
          adjustment={adjustment}
          close={close}
          onDelete={this.deleteAdjustment}
          rootPath={rootPath}
        />
        <Callout ref={this.createCalloutRef} />
      </Fragment>
    );
  }
}

export default stripesConnect(SettingsAdjustmentsViewContainer);
