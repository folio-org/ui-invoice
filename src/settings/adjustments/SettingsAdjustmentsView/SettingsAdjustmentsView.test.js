import React from 'react';
import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';

import {
  Paneset,
  HasCommand,
} from '@folio/stripes/components';

import { adjustment, history } from '../../../../test/jest/fixtures';

import SettingsAdjustmentsView from './SettingsAdjustmentsView';

jest.mock('@folio/stripes-components/lib/Commander', () => ({
  HasCommand: jest.fn(({ children }) => <div>{children}</div>),
}));
jest.mock('@folio/stripes-components/lib/NoValue', () => {
  return () => <span>-</span>;
});

const mockPush = jest.fn();
const defaultProps = {
  adjustment: {
    metadata: {},
    title: 'adjustment title',
    adjustment,
  },
  close: jest.fn(),
  onDelete: jest.fn(),
  rootPath: '/path',
  history: { ...history, push: mockPush },
  stripes: { hasPerm: jest.fn().mockReturnValue(true) },
};
const renderSettingsAdjustmentsView = (props = defaultProps) => render(
  <Paneset>
    <SettingsAdjustmentsView {...props} />
  </Paneset>,
  { wrapper: MemoryRouter },
);

describe('SettingsAdjustmentsView', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure', async () => {
    const { asFragment } = renderSettingsAdjustmentsView();

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Actions', () => {
    it('should open confirmation modal when delete action is pressed', () => {
      renderSettingsAdjustmentsView();

      user.click(screen.getByTestId('adjustment-delete'));

      expect(screen.getByText('ui-invoice.settings.adjustments.confirmDelete.message')).toBeDefined();
    });

    it('should call onDelete prop when delete is confirmed', () => {
      const onDelete = jest.fn();

      renderSettingsAdjustmentsView({ ...defaultProps, onDelete });

      user.click(screen.getByTestId('adjustment-delete'));
      user.click(screen.getByText('ui-invoice.settings.adjustments.confirmDelete.confirmLabel'));

      expect(onDelete).toHaveBeenCalled();
    });
  });

  describe('SettingsAdjustmentsView shortcuts', () => {
    beforeEach(() => {
      HasCommand.mockClear();
      defaultProps.close.mockClear();
    });

    it('should call close when cancel shortcut is called', () => {
      renderSettingsAdjustmentsView();

      HasCommand.mock.calls[0][0].commands.find(c => c.name === 'cancel').handler();

      expect(defaultProps.close).toHaveBeenCalled();
    });

    it('should navigate to edit form when edit shortcut is called', () => {
      mockPush.mockClear();

      renderSettingsAdjustmentsView();

      HasCommand.mock.calls[0][0].commands.find(c => c.name === 'edit').handler();

      expect(mockPush).toHaveBeenCalledWith(`${defaultProps.rootPath}/${defaultProps.adjustment.id}/edit`);
    });
  });
});
