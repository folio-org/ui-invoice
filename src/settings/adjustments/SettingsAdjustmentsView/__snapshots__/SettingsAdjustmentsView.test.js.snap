// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsAdjustmentsView should render correct structure 1`] = `
<DocumentFragment>
  <div
    class="paneset  "
  >
    <div
      aria-label="Adjustment details"
      class="LayerRoot"
      role="dialog"
      tabindex="-1"
    >
      <div>
        <section
          class="pane focusIndicator"
          id="invoice-settings-adjustment-view"
        >
          <div
            aria-labelledby="paneHeaderinvoice-settings-adjustment-view-pane-title"
            class="paneHeader"
            data-test-pane-header="true"
            id="paneHeaderinvoice-settings-adjustment-view"
            role="region"
            tabindex="-1"
          >
            <div
              class="paneHeaderButtonsArea first"
            >
              <div
                class="paneMenu"
              >
                <button
                  aria-label="stripes-components.closeItem"
                  class="iconButton medium paneHeaderCloseIcon PaneHeaderIconButton paneHeaderIconButton"
                  data-test-pane-header-dismiss-button="true"
                  icon="times"
                  type="button"
                >
                  <span
                    class="iconButtonInner mediumOffset PaneHeaderIconButton__inner"
                    data-test-pane-header-dismiss-button="true"
                  >
                    <span>
                      Icon
                    </span>
                  </span>
                </button>
              </div>
            </div>
            <div
              class="paneHeaderCenter"
            >
              <div
                class="paneHeaderCenterInner"
                id="paneHeaderinvoice-settings-adjustment-view-pane-title"
              >
                <h2
                  class="paneTitle"
                  data-test-pane-header-title="true"
                >
                  <span
                    class="paneTitleLabel"
                  >
                    adjustment title
                  </span>
                </h2>
              </div>
            </div>
            <div
              class="paneHeaderButtonsArea last"
            >
              <div
                class="paneMenu"
              >
                <div
                  class="dropdown"
                  data-pane-header-actions-dropdown="true"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="button primary marginBottom0 actionMenuToggle"
                    data-test-pane-header-actions-button="true"
                    type="button"
                  >
                    <span
                      class="inner"
                    >
                      <span>
                        Icon
                      </span>
                    </span>
                  </button>
                  <div
                    style="display: none;"
                  >
                    <div
                      class="DropdownMenu"
                      style="left: 0px; display: block;"
                    >
                      <div
                        data-test-view-adjustment-actions="true"
                      >
                        <a
                          class="button dropdownItem"
                          data-test-view-adjustment-action-edit="true"
                          href="/path/undefined/edit"
                          role="button"
                        >
                          <span>
                            Icon
                          </span>
                        </a>
                        <button
                          class="button dropdownItem"
                          data-test-view-adjustment-action-delete="true"
                          data-testid="adjustment-delete"
                          type="button"
                        >
                          <span
                            class="inner"
                          >
                            <span>
                              Icon
                            </span>
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="paneContent hasPadding"
            id="invoice-settings-adjustment-view-content"
            tabindex="-1"
          >
            <div
              class="row center-xs"
            >
              <div
                class="col-xs-12 col-md-8"
              >
                <div
                  class="row start-xs"
                >
                  <div
                    class="col-xs-12"
                  >
                    <div>
                      ViewMetaData
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-description="true"
                  >
                    <div
                      class="kvRoot"
                    >
                      <div
                        class="kvLabel"
                      >
                        ui-invoice.settings.adjustments.description
                      </div>
                      <div
                        class="kvValue"
                        data-test-kv-value="true"
                      >
                        <span>
                          -
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-type="true"
                  >
                    <div
                      class="kvRoot"
                    >
                      <div
                        class="kvLabel"
                      >
                        ui-invoice.settings.adjustments.type
                      </div>
                      <div
                        class="kvValue"
                        data-test-kv-value="true"
                      >
                        Amount
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-always-show="true"
                  >
                    <div
                      class="checkbox inline vertical disabled"
                      data-test-checkbox="true"
                    >
                      <label
                        class="label label"
                        for="checkbox-2"
                      >
                        <span
                          class="labelText"
                        >
                          ui-invoice.settings.adjustments.alwaysShow
                        </span>
                        <span
                          class="inner disabledLabel"
                        >
                          <input
                            aria-invalid="false"
                            class="input"
                            disabled=""
                            id="checkbox-2"
                            type="checkbox"
                            value=""
                          />
                          <span
                            class="checkboxIcon"
                          >
                            <svg
                              height="10"
                              viewBox="0 0 14 14"
                              width="10"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M10.395 4.617L5.63 9.383 3.605 7.358"
                                fill="none"
                                stroke-linecap="round"
                                stroke-linejoin="round"
                                stroke-miterlimit="10"
                                stroke-width="1.5"
                              />
                            </svg>
                          </span>
                        </span>
                      </label>
                      <div
                        role="alert"
                      />
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-default-amount="true"
                  >
                    <div
                      class="kvRoot"
                    >
                      <div
                        class="kvLabel"
                      >
                        ui-invoice.settings.adjustments.defaultAmount
                      </div>
                      <div
                        class="kvValue"
                        data-test-kv-value="true"
                      >
                        <span>
                          -
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-prorate="true"
                  >
                    <div
                      class="kvRoot"
                    >
                      <div
                        class="kvLabel"
                      >
                        ui-invoice.settings.adjustments.prorate
                      </div>
                      <div
                        class="kvValue"
                        data-test-kv-value="true"
                      >
                        <span>
                          -
                        </span>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-relation-to-total="true"
                  >
                    <div
                      class="kvRoot"
                    >
                      <div
                        class="kvLabel"
                      >
                        ui-invoice.settings.adjustments.relationToTotal
                      </div>
                      <div
                        class="kvValue"
                        data-test-kv-value="true"
                      >
                        In addition to
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-xs-3"
                    data-test-export-to-accounting="true"
                  >
                    <div
                      class="checkbox inline vertical disabled"
                      data-test-checkbox="true"
                    >
                      <label
                        class="label label"
                        for="checkbox-3"
                      >
                        <span
                          class="labelText"
                        >
                          ui-invoice.settings.adjustments.exportToAccounting
                        </span>
                        <span
                          class="inner disabledLabel"
                        >
                          <input
                            aria-invalid="false"
                            class="input"
                            disabled=""
                            id="checkbox-3"
                            type="checkbox"
                            value=""
                          />
                          <span
                            class="checkboxIcon"
                          >
                            <svg
                              height="10"
                              viewBox="0 0 14 14"
                              width="10"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M10.395 4.617L5.63 9.383 3.605 7.358"
                                fill="none"
                                stroke-linecap="round"
                                stroke-linejoin="round"
                                stroke-miterlimit="10"
                                stroke-width="1.5"
                              />
                            </svg>
                          </span>
                        </span>
                      </label>
                      <div
                        role="alert"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>
  <div
    class="container"
  >
    <div
      class="cursorHover"
      hidden=""
      style="left: -10px;"
    >
      <div
        class="cursor"
      >
         
      </div>
    </div>
  </div>
</DocumentFragment>
`;
